src/Microsoft.Crank.Agent/Controllers/JobsController.cs

```diff
@@ -332,6 +332,64 @@ public async Task<IActionResult> UploadAttachment(int id)
             return Ok();
         }
 
+        [HttpPost("{id}/attachment/zip")]
+        [RequestSizeLimit(10_000_000_000)]
+        [RequestFormLimits(MultipartBodyLengthLimit = 10_000_000_000)]
+        public async Task<IActionResult> UploadAttachmentZip(int id)
+        {
+            // Because uploaded files can be big, the client is supposed to keep
+            // pinging the service to keep the job alive during the Initialize state.
+
+            var destinationFilename = Request.Headers["destinationFilename"].ToString();
+
+            Log.Info($"Uploading archive: {destinationFilename}");
+
+            var job = _jobs.Find(id);
+
+            if (job == null)
+            {
+                return NotFound("Job doesn't exist anymore");
+            }
+
+            if (job.State != JobState.Initializing)
+            {
+                Log.Info($"Attachment rejected, job is {job.State}");
+                return StatusCode(500, $"The job can't accept attachment as its state is {job.State}");
+            }
+
+            var tempFilename = Path.GetTempFileName() + ".zip";
+
+            await SaveBodyAsync(tempFilename);
+
+            job.LastDriverCommunicationUtc = DateTime.UtcNow;
+
+            var destinationTempFilename = Path.GetFullPath(Path.GetRandomFileName(), Path.GetTempPath());
+            
+            // Extract the zip file in a temporary folder
+            ZipFile.ExtractToDirectory(tempFilename, destinationTempFilename);
+
+            job.LastDriverCommunicationUtc = DateTime.UtcNow;
+
+            System.IO.File.Delete(tempFilename);
```

Comment:
What if some unexpected error happens above? I'd suggest wrapping lines 362-371 in a try - finally block and moving the `Delete` into a finally block.

Reply:
Done